File: drivers/dart/lib/src/options.dart
import 'dart:io' show SecurityContext;

/// Top-level options for `connect()`. Most users won't need to fill this
/// out — the URI carries the same information.
class ConnectOptions {
  const ConnectOptions({
    this.clientName = 'reddb-dart/0.1.0',
    this.timeout = const Duration(seconds: 30),
    this.token,
    this.username,
    this.password,
    this.tls,
  });

  final String clientName;
  final Duration timeout;

  /// Static bearer token. Wins over `username`/`password` (no auto-login).
  final String? token;

  /// HTTP basic-style credentials. When set the HTTP transport calls
  /// `/auth/login` on connect to exchange them for a bearer token.
  final String? username;
  final String? password;

  /// Optional TLS configuration. When supplied for a `red://` URI the
  /// driver promotes the connection to TLS; for `reds://` it is merged
  /// with the URI's defaults.
  final TlsOptions? tls;

  ConnectOptions copyWith({
    String? clientName,
    Duration? timeout,
    String? token,
    String? username,
    String? password,
    TlsOptions? tls,
  }) {
    return ConnectOptions(
      clientName: clientName ?? this.clientName,
      timeout: timeout ?? this.timeout,
      token: token ?? this.token,
      username: username ?? this.username,
      password: password ?? this.password,
      tls: tls ?? this.tls,
    );
  }
}

/// TLS configuration for the RedWire transport.
class TlsOptions {
  const TlsOptions({
    this.context,
    this.servername,
    this.allowInsecure = false,
    this.alpnProtocols = const ['redwire/1'],
  });

  /// Override the default trust store. `null` uses the platform default.
  final SecurityContext? context;

  /// SNI override. Defaults to the connect host.
  final String? servername;

  /// When true, accept self-signed / mismatched certs. Dev-only.
  final bool allowInsecure;

  /// ALPN tokens offered to the server. Default: `['redwire/1']`.
  final List<String> alpnProtocols;
}

File: drivers/java/src/main/java/dev/reddb/Options.java
package dev.reddb;

import java.time.Duration;

/**
 * Immutable bag of optional connection knobs. Pass to
 * {@link Reddb#connect(java.net.URI, Options)}. Use the static
 * builder when you need to override a default.
 */
public final class Options {
    /** Default — empty, no overrides. */
    public static final Options DEFAULTS = new Options(null, null, null, null, null, Duration.ofSeconds(30));

    private final String username;
    private final String password;
    private final String token;
    private final String apiKey;
    private final String clientName;
    private final Duration timeout;

    private Options(String username, String password, String token, String apiKey,
                    String clientName, Duration timeout) {
        this.username = username;
        this.password = password;
        this.token = token;
        this.apiKey = apiKey;
        this.clientName = clientName;
        this.timeout = timeout == null ? Duration.ofSeconds(30) : timeout;
    }

    public String username() { return username; }
    public String password() { return password; }
    public String token() { return token; }
    public String apiKey() { return apiKey; }
    public String clientName() { return clientName; }
    public Duration timeout() { return timeout; }

    public static Builder builder() { return new Builder(); }

    public static final class Builder {
        private String username;
        private String password;
        private String token;
        private String apiKey;
        private String clientName;
        private Duration timeout = Duration.ofSeconds(30);

        public Builder username(String v) { this.username = v; return this; }
        public Builder password(String v) { this.password = v; return this; }
        public Builder token(String v) { this.token = v; return this; }
        public Builder apiKey(String v) { this.apiKey = v; return this; }
        public Builder clientName(String v) { this.clientName = v; return this; }
        public Builder timeout(Duration v) { this.timeout = v; return this; }

        public Options build() {
            return new Options(username, password, token, apiKey, clientName, timeout);
        }
    }
}
